### ISSUE TITLE
Extended resources via kubelet configuration

### ISSUE THREAD

**Issue by sethp-nr** [kubernetes/kubernetes#82302]
**What would you like to be added**:

A kubeadm-compatible mechanism for managing extended resources. A possible approach can be found in #81858.

This would (in my mind) establish parity around the "node feature advertisement" that node labels offer for extended resources.

**Why is this needed**:

In short, the existing mechanisms for managing extended resources don't well cover our use-case. Those mechanisms are:

- A controller that watches nodes and updates the capacity as appropriate
    - https://github.com/kubernetes-sigs/node-feature-discovery looks like a neat general-purpose implementation
- A novel device controller that uses the existing device controller mechanism to advertise resources with a no-op "attach/detach"
- A bootstrap script (or manual process) that includes a "once the node appears, run this to apply the right capacity"

I have a handful of static extended resources, along with node labels and taints, that I would like to apply "at the same time" (read: configured by a similar mechanism) as node labels and taints.

**The use-case**:

Without further ado, the X part of this [X-Y problem](https://en.wikipedia.org/wiki/XY_problem).

We're running a large, stateful system on a kubernetes cluster managed by Cluster API (CRDs for the cluster itself). Those CRDs expose a kubeadm config for each node (via the MachineDeployment) that we've used to express certain node labels and taints for various different scaling dimensions. We taint the nodes to ensure nothing besides the specific workload lands there, and so scaling looks like 1) scale the number of nodes by increasing the MachineDeployment's `replicas` and 2) define Pods sufficient to make use of the new capacity.

In order to ensure the number of workers we expect are scheduled to each node, we define a pod that consumes an extended resource. Sometimes that's mediated by volume availability (if the local volume provisioner exposes two volumes, then only two workers will be scheduled) and the existing volume device plugin works well for exposing the resource, but in other cases it's a virtual `num-workers` resource that serves no other purpose than to limit the number of cohabitants.

Another complicating factor is that these workers deploy by "hotswap" – we have to land the replacement Pod on the same Node _before_ deleting the replaced Pod. What we're looking to achieve with this scheme is a managed handoff between the two Pods, which in our case implies shared CPU resources (the total volume of work doesn't change, just the ratio between the Pods). As I'm not aware of a way to resize the CPU request/limit "just in time" or to explicitly share resources, we use a Pod that does nothing besides "stake out" one of the virtual resources and then affine the worker pod (and its replacements) to it. None of the worker pods set CPU requests or limits, allowing us to manage CPU sharing outside of kubernetes without giving up the volume / disk gc / memory / etc. goodness the kubelet provides.

So, to sum up, we're looking to deploy exactly `num-workers + num-hotswaps` Pods to a Node, where `num-hotswaps` is zero unless one or more of the workers is being deployed. To solve that, we're actually using two pods per worker: one that has a no-op "pause" container that consumes only a `num-worker` extended resource, and the actual worker pods (including replacements) define a required affinity to the first "pause" pod. We want to use a node-size-specific `kubeadm` config to define that number of workers.

/sig node

**Comment by sethp-nr** [comment]
cc @dashpole @dchen1107 

I didn't catch the name of the other person asking for more details on the SIG node meeting, would one of you loop them in?

**Comment by sethp-nr** [comment]
The ability to set extended resource capacity via the kubelet without requiring a device driver is still desirable to us, as our current daemonset-based workaround has a few usability rough edges we'd like to address.
